### PR TITLE
Payrate Updates

### PR DESCRIPTION
**Pull request by ellisbrian-fresh**
**Ticket**:  [Payrate Update Process](https://app.asana.com/0/1200418423170087/1204925811708944)

- Added functionality to update the payrate for employees in Intacct

- Updated request formatting to allow for objects without keys (necessary for updating the payrate and other future functions)

**Testing**
Used a Tap-CSV to Target-Intacct pipeline to upload data to the Intacct sandbox

_CSV File_
image

_Result in Intacct_
image

**Notes**
- Date format can be any variation of Month/Day/Year (ex: 04/4/23, 04/04/2023, etc), which is the Intacct standard and format indicated by Zoe
- The salaryrate column is not used for our purposes, but is implemented for potential future or outside use

### CODE REVIEW

File: src/target_intacct/employee_rate.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                        "day": day
+                    },
+                    "billingrate": row.get("billingrate", ""),
+                    "salaryrate": row.get("salaryrate", ""),
+                
+            }
+            intacct_client.post_employee_rate(entry)
+        else:
+            raise Exception(
+            "Missing Required employeeid Column. At least one employee id is required to upload a employee payrate"
```
Comment (hulet) [inline src/target_intacct/employee_rate.py:57]: If I'm reading this right this exception can trigger if there is an `employeeid` column that does contain a value but that value isn't an employee id in Intacct? "Missing Required employeeid Column" makes me think the column doesn't exist at all. "At least one employee id is required to upload a employee payrate" - can this trigger even when one employee ID is provided (but that employee isn't in Intacct)? If so perhaps something like "Invalid employeeid - Employee ID does not exist in Intacct"?

File: src/target_intacct/client.py [hunk 2]
```diff
@@ -229,10 +229,18 @@ def format_and_send_request(self, data: Dict) -> Union[List, Dict]:
         Returns:
             A response from the _post_request (dict).
         """
-
         key = next(iter(data))
-        object_type = data[key]["object"]
 
+        if use_key:
+            object_type = data[key]["object"]
+
```
Comment (hulet) [inline src/target_intacct/client.py:236]: Consider deleting so the formatting of the two assignment blocks is identical, which makes the two branches easier to compare.

File: src/target_intacct/employee_rate.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                    "ratestartdate": {
+                        "year": year,
+                        "month": month,
+                        "day": day
+                    },
+                    "billingrate": row.get("billingrate", ""),
+                    "salaryrate": row.get("salaryrate", ""),
+                
+            }
+            intacct_client.post_employee_rate(entry)
```
Comment (hulet) [inline src/target_intacct/employee_rate.py:54]: Will this make one Intacct API call per input row? Is there an option to batch these into fewer requests?

File: src/target_intacct/employee_rate.py (hunk 1 above)
Comment (ellisbrian-fresh) [reply src/target_intacct/employee_rate.py:57]: Yep, thanks for catching that, that's something I forgot to update. That's exactly the purpose of it

File: src/target_intacct/employee_rate.py (hunk 3 above)
Comment (ellisbrian-fresh) [reply src/target_intacct/employee_rate.py:54]: hmmm good call, there should be

**Review by hulet — APPROVED** [review]

**Pull request merged**